clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -313,6 +317,13 @@ private KafkaProducer(ProducerConfig config, Serializer<K> keySerializer, Serial
                 config.ignore(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG);
                 this.valueSerializer = valueSerializer;
             }
+
+            // load interceptors and make sure they get clientId
+            userProvidedConfigs.put(ProducerConfig.CLIENT_ID_CONFIG, clientId);
+            List<ProducerInterceptor<K, V>> interceptorList = (List) (new ProducerConfig(userProvidedConfigs)).getConfiguredInstances(ProducerConfig.INTERCEPTOR_CLASSES_CONFIG,
+                    ProducerInterceptor.class);
+            this.interceptors = new ProducerInterceptors<>(interceptorList);
```

Comment:
Yes, that may be better. I am also allocating an extra callback object on each record send. I will do this change in my next commit and see how that looks.